x/sync/network_server.go

```diff
@@ -183,30 +256,91 @@ func (s *NetworkServer) HandleRangeProofRequest(
+		bytesLimit = constants.DefaultMaxMessageSize
 	}
+	for keyLimit > 0 {
+		rangeProof, err := s.db.GetRangeProofAtRoot(ctx, req.Root, req.Start, req.End, int(keyLimit))
+		if err != nil {
+			// handle expected errors so clients cannot cause servers to spam warning logs.
+			if errors.Is(err, merkledb.ErrRootIDNotPresent) {
+				s.log.Debug(
+					"dropping invalid range proof request",
+					zap.Stringer("nodeID", nodeID),
+					zap.Uint32("requestID", requestID),
+					zap.Stringer("req", req),
+					zap.Error(err),
+				)
+				return nil // dropping request
+			}
+			return err
+		}
+
+		proofBytes, err := merkledb.Codec.EncodeRangeProof(Version, rangeProof)
+		if err != nil {
+			return err
+		}
+		if len(proofBytes) < bytesLimit {
+			return s.appSender.SendAppResponse(ctx, nodeID, requestID, proofBytes)
+		}
+		// the proof size was too large, try to shrink it
+
+		// ensure that the new limit is always smaller
+		keyLimit = uint16(len(rangeProof.KeyValues) / 2)
+
+		// estimate the bytes of the start and end proof to ensure that everything will fit into the bytesLimit
+		bytesEstimate := getBytesEstimateOfProofNodes(rangeProof.StartProof)
 
-	rangeProof, err := s.db.GetRangeProofAtRoot(ctx, req.Root, req.Start, req.End, int(limit))
-	if err != nil {
-		// handle expected errors so clients cannot cause servers to spam warning logs.
-		if errors.Is(err, merkledb.ErrRootIDNotPresent) {
-			s.log.Debug(
-				"dropping invalid range proof request",
+		// just the start proof is too large, so a proof is impossible
+		if bytesEstimate > int(req.BytesLimit) {
+			// errors are fatal, so log for the moment
+			s.log.Warn(
+				"cannot generate a proof within bytes limit",
 				zap.Stringer("nodeID", nodeID),
 				zap.Uint32("requestID", requestID),
 				zap.Stringer("req", req),
-				zap.Error(err),
+				zap.Error(ErrMinProofSizeIsTooLarge),
 			)
-			return nil // dropping request
+			return nil
+		}
+
+		bytesEstimate += getBytesEstimateOfProofNodes(rangeProof.EndProof)
+
+		// shrink more if the early keys are extremely large
+		for keyIndex := uint16(1); keyIndex < keyLimit; keyIndex++ {
+			nextKV := rangeProof.KeyValues[keyIndex]
+			kvEstBytes := len(nextKV.Key) + len(nextKV.Value) + 2*binary.MaxVarintLen64
+
+			if bytesEstimate+kvEstBytes > bytesLimit {
+				// adding the current KV would put the size over the limit
+				// so only return up to the keyIndex number of keys
+				keyLimit = keyIndex
+				break
+			}
+			bytesEstimate += kvEstBytes
 		}
-		return err
 	}
+	// errors are fatal, so log for the moment
+	s.log.Warn(
+		"cannot generate a proof within bytes limit",
+		zap.Stringer("nodeID", nodeID),
+		zap.Uint32("requestID", requestID),
+		zap.Stringer("req", req),
+		zap.Error(ErrMinProofSizeIsTooLarge),
+	)
+	return nil
+}
 
-	proofBytes, err := merkledb.Codec.EncodeRangeProof(Version, rangeProof)
-	if err != nil {
-		return err
+func getBytesEstimateOfProofNodes(proofNodes []merkledb.ProofNode) int {
```

Comment:
I think this should probably live inside the codec -- we should try to keep the sync stuff as agnostic to the codec as possible